pytorch/wavenet.py

```diff
@@ -0,0 +1,214 @@
+            if i < len(self.res_layers):
+                res_acts = self.res_layers[i](acts)
+            forward_input = res_acts + forward_input
+
+            if i == 0:
+                output = self.skip_layers[i](acts)
+            else:
+                output = self.skip_layers[i](acts) + output
+
+        output = torch.nn.functional.relu(output, True)
+        output = self.conv_out(output)
+        output = torch.nn.functional.relu(output, True)
+        output = self.conv_end(output)
+
+        # Remove last probabilities because they've seen all the data
+        last = output[:, :, -1]
+        last = last.unsqueeze(2)
+        output = output[:, :, :-1]
+
+        # Replace probability for first value with 0's because we don't know
+        first = last * 0.0
+        output = torch.cat((first, output), dim=2)
+
+        return output
+
+    def make_nv_wavenet(self):
+        """
+        Returns a dictionary with tensors ready for nv_wavenet wrapper
+        """
+        model = {}
+        # We're not using a convolution to start to this does nothing
+        model["embedding_prev"] = torch.cuda.FloatTensor(self.n_out_channels,
+                                              self.n_residual_channels).fill_(0.0)
+
+        model["embedding_curr"] = self.embed.weight.data
+        model["conv_out_weight"] = self.conv_out.conv.weight.data
+        model["conv_end_weight"] = self.conv_end.conv.weight.data
+        
+        dilate_weights = []
+        dilate_biases = []
+        for layer in self.dilate_layers:
+            dilate_weights.append(layer.conv.weight.data)
+            dilate_biases.append(layer.conv.bias.data)
+        model["dilate_weights"] = dilate_weights
+        model["dilate_biases"] = dilate_biases
+       
+        model["max_dilation"] = self.max_dilation
+
+        res_weights = []
+        res_biases = []
+        for layer in self.res_layers:
+            res_weights.append(layer.conv.weight.data)
+            res_biases.append(layer.conv.bias.data)
+        model["res_weights"] = res_weights
+        model["res_biases"] = res_biases
+        
+        skip_weights = []
+        skip_biases = []
+        for layer in self.skip_layers:
+            skip_weights.append(layer.conv.weight.data)
+            skip_biases.append(layer.conv.bias.data)
+        model["skip_weights"] = skip_weights
+        model["skip_biases"] = skip_biases
+        
+        model["use_embed_tanh"] = False
+    
+        return model
+
+    def get_cond_input(self, features):
+        """
+        Takes in features and gets the 2*R x batch x # layers x samples tensor
+        """
+        # TODO(rcosta): trim conv artifacts. mauybe pad spec to kernel multiple
+        cond_input = self.upsample(features)
+        time_cutoff = self.upsample.kernel_size[0] - self.upsample.stride[0]
+        cond_input = cond_input[:, :, :-time_cutoff]
+        cond_acts = [layer.conv(cond_input).data for layer in self.cond_layers]
```

Comment:
Use a 2d Conv instead, for performance.

Reply:
https://discuss.pytorch.org/t/can-this-be-achieved/18740/4

Reply:
I made it one big 1D conv instead.  I think it trains a little faster now, but should definitely help the latency on inference.